src/librustc/hir/lowering.rs

```diff
@@ -143,8 +143,13 @@ pub struct LoweringContext<'a> {
 }
 
 pub trait Resolver {
-    /// Resolve a hir path generated by the lowerer when expanding `for`, `if let`, etc.
-    fn resolve_hir_path(&mut self, path: &mut hir::Path, is_value: bool);
+    /// Resolve a path generated by the lowerer when expanding `for`, `if let`, etc.
+    fn resolve_hir_path(
+        &mut self,
+        path: &ast::Path,
+        args: Option<P<hir::GenericArgs>>,
```

Comment:
Looks like `args` travel to resolve through all these interfaces not to be resolved, but only to be put into the last segment of the path.
I'd personally keep the old signature form `resolve_hir_path` and made a `lowering.rs`-local helper adding `args` to the last segment.